src/main/java/org/opensearch/knn/index/vectorvalues/KNNVectorValuesIterator.java

```
@@ -0,0 +1,157 @@
+     */
+    int nextDoc() throws IOException;
+
+    /**
+     * Return a {@link DocIdSetIterator}
+     * @return {@link DocIdSetIterator}
+     */
+    DocIdSetIterator getDocIdSetIterator();
+
+    /**
+     * Total number of live doc which will the iterator will iterate upon.
+     * @return long: total number of live docs
+     */
+    long liveDocs();
+
+    /**
+     * A DocIdsIteratorValues provides a common iteration logic for all Values that implements
+     * {@link DocIdSetIterator} interface. Example: {@link BinaryDocValues}, {@link FloatVectorValues} etc.
+     */
+    class DocIdsIteratorValues implements KNNVectorValuesIterator {
+        protected DocIdSetIterator docIdSetIterator;
+
+        DocIdsIteratorValues(@NonNull final DocIdSetIterator docIdSetIterator) {
+            this.docIdSetIterator = docIdSetIterator;
+        }
+
+        @Override
+        public int docId() {
+            return docIdSetIterator.docID();
+        }
+
+        @Override
+        public int advance(int docId) throws IOException {
+            return docIdSetIterator.advance(docId);
+        }
+
+        @Override
+        public int nextDoc() throws IOException {
+            return docIdSetIterator.nextDoc();
+        }
+
+        @Override
+        public DocIdSetIterator getDocIdSetIterator() {
+            return docIdSetIterator;
+        }
+
+        @Override
+        public long liveDocs() {
+            if (docIdSetIterator instanceof BinaryDocValues) {
+                return KNNCodecUtil.getTotalLiveDocsCount((BinaryDocValues) docIdSetIterator);
+            } else if (docIdSetIterator instanceof FloatVectorValues || docIdSetIterator instanceof ByteVectorValues) {
+                return docIdSetIterator.cost();
+            }
+            throw new IllegalArgumentException(
+                "DocIdSetIterator present is not of valid type. Valid types are: BinaryDocValues, FloatVectorValues and ByteVectorValues"
```

I think the validation should be in the constructor. This might force us to validate in other methods or logic being added in this class. The earlier the validation the better. Let me know what you think

---

I agree that validation can be moved to the constructor, but even in that case this exception will be present that is curse of `if/else`. we need to have a else condition. Lets say even if I move this logic via Map<itr, functional interface> I would still need to throw the exception there as the output of map can be null for an iterator which is not present in map.

---

Added the validations in constructor too.